phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -409,91 +413,220 @@ private void populatePendingRows(BatchMutateContext context) {
-                      mutationsMap.put(row, stored);
+  /**
+   * * Merges the mutations on the same row
+   */
+  private Collection<? extends Mutation> mergeMutations(ObserverContext<RegionCoprocessorEnvironment> c,
+                                                        MiniBatchOperationInProgress<Mutation> miniBatchOp,
+                                                        BatchMutateContext context,
+                                                        long now, boolean rebuild) throws IOException {
+      if (!rebuild) {
+          if (context.rowsToLock.size() == 0) {
+              return null;
+          }
+          Map<ImmutableBytesPtr, Integer> pendingPuts = new HashMap<>(miniBatchOp.size());
+          mergePendingPutMutations(miniBatchOp, pendingPuts, now);
+          getCurrentRowStates(c, context);
+          context.mergedRowStates = new HashMap<ImmutableBytesPtr, Put>(miniBatchOp.size());
+          // Merge pending put mutations with current row states into new merged states
+          for (Integer i = 0; i < miniBatchOp.size(); i++) {
+              if (miniBatchOp.getOperationStatus(i) == IGNORE || miniBatchOp.getOperationStatus(i) == NOWRITE) {
+                  continue;
+              }
+              Mutation m = miniBatchOp.getOperation(i);
+              if (this.builder.isEnabled(m)) {
+                  if (m instanceof Put) {
+                      ImmutableBytesPtr rowKeyPtr = new ImmutableBytesPtr(m.getRow());
+                      Put currentRow = context.currentRowStates.get(rowKeyPtr);
+                      Put mergedRow;
+                      if (currentRow != null) {
+                          mergedRow = mergeNew((Put) m, currentRow);
+                      }
+                      else {
+                          mergedRow = new Put((Put)m);
+                      }
+                      context.mergedRowStates.put(rowKeyPtr, mergedRow);
                   }
-                  stored.addAll(m);
-              } else {
-                  originalMutations.add(m);
               }
           }
+          // Apply delete mutations on the pending put mutations and merged row states
+          applyPendingDeleteMutations(miniBatchOp, context, pendingPuts);
       }
 
-      if (copyMutations || replayWrite != null) {
-          mutations = IndexManagementUtil.flattenMutationsByTimestamp(mutations);
+      Collection<Mutation> pendingMutations = Lists.newArrayListWithExpectedSize(miniBatchOp.size());
+      // Add put mutations into the collection first. This ordering of puts first and then deletes is important for
+      // obtaining correct index updates
+      for (int i = 0; i < miniBatchOp.size(); i++) {
+          Mutation m = miniBatchOp.getOperation(i);
+          if (m instanceof Put) {
+              if (miniBatchOp.getOperationStatus(i) != IGNORE &&
+                      miniBatchOp.getOperationStatus(i) != NOWRITE && this.builder.isEnabled(m)) {
```

Comment:
can extract this to a helper method called by both puts and deletes.

Reply:
ok